platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/ResolutionCandidateAssessor.java

```diff
@@ -72,11 +73,13 @@ public List<AssessedCandidate> assessResolvedVariants(List<? extends ResolvedVar
             .collect(Collectors.toList());
     }
 
-    public List<AssessedCandidate> assessResolvedVariantStates(List<? extends VariantGraphResolveState> variantStates) {
+    public List<AssessedCandidate> assessResolvedVariantStates(List<? extends VariantGraphResolveState> variantStates, ImmutableCapability defaultCapabilityForComponent) {
         return variantStates.stream()
             .map(VariantGraphResolveState::getMetadata)
-            .map(variant -> assessCandidate(variant.getName(), variant.getCapabilities(), variant.getAttributes().asImmutable()))
-            .sorted(Comparator.comparing(AssessedCandidate::getDisplayName))
+            .map(variant -> {
+                ImmutableCapabilities capabilities = new ImmutableCapabilities(variant.getCapabilities().asSet(), defaultCapabilityForComponent);
```

Comment:
I am not really a fan of adding this new `ImmutableCapabilities` constructor -- especially since we need to unwrap an existing `ImmutableCapabilities` instance using `asSet` just to call this new constructor. 

I would much prefer we add an `orElse` method to `ImmutableCapabilities` that returns `this` if the immutable capabilities is non-empty, or the parameter if the capabilities are empty. 

Doing so would make this code read like:
```
ImmutableCapabilities capabilities = variant.getCapabilities().orElse(defaultCapabilityForComponent)
```

Reply:
I like it, I'll update.